rfcs/0000-user-defined-aggregate-functions.md

```diff
@@ -0,0 +1,269 @@
+    def deserialize(self, serialized: bytes) -> State:
+        # default implementation
+        return pickle.loads(serialized)
+```
+
+Reference: [UDAF Python API in Flink](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/python/table/udfs/python_udfs/#aggregate-functions)
+
+### Java API
+
+The Java API is similar to Python. One of the differences is that users don't need to specify the input and output types, because we can infer them from the function signatures.
+
+```java
+import java.io.Serializable;
+import com.risingwave.functions.AggregateFunction;
+
+// Mutable accumulator for the aggregate function.
+// This class should be Serializable.
+public class WeightedAvgAccumulator implements Serializable {
+    public long sum = 0;
+    public int count = 0;
+}
+
+// The aggregate function is defined as a class, which implements the `AggregateFunction` interface.
+public class WeightedAvg implements AggregateFunction {
+    // Create a new accumulator.
+    public WeightedAvgAccumulator createAccumulator() {
+        return new WeightedAvgAccumulator();
+    }
+
+    // Get the aggregate result.
+    // The result type is inferred from the signature. (BIGINT)
+    // If a Java type can not infer to an unique SQL type, it should be annotated with `@DataTypeHint`.
+    public Long getValue(WeightedAvgAccumulator acc) {
+        if (acc.count == 0) {
+            return null;
+        } else {
+            return acc.sum / acc.count;
+        }
+    }
+
+    // Accumulate a row to the accumulator.
+    // The input types are inferred from the signatures. (BIGINT, INT)
+    // If a Java type can not infer to an unique SQL type, it should be annotated with `@DataTypeHint`.
+    public void accumulate(WeightedAvgAccumulator acc, long iValue, int iWeight) {
+        acc.sum += iValue * iWeight;
+        acc.count += iWeight;
+    }
+
+    // Retract a row from the accumulator. (optional)
+    // The function signature should match `accumulate`.
+    public void retract(WeightedAvgAccumulator acc, long iValue, int iWeight) {
+        acc.sum -= iValue * iWeight;
+        acc.count -= iWeight;
+    }
+
+    // Merge the accumulator with another one. (optional)
+    public void merge(WeightedAvgAccumulator acc, WeightedAvgAccumulator other) {
+        acc.count += other.count;
+        acc.sum += other.sum;
+    }
+}
+```
+
+Reference: [UDAF Java API in Flink](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/table/functions/udfs/#aggregate-functions)
+
+### SQL API
+
+After setting up the UDF server, users can register the function through `create aggregate` in RisingWave.
+
+```sql
+create aggregate weighted_avg(value bigint, weight int) returns bigint
+as 'weighted_avg' using link 'http://localhost:8815';
+```
+
+The full syntax is:
+
+```sql
+CREATE [ OR REPLACE ] AGGREGATE name ( [ argname ] arg_data_type [ , ... ] )
+[ RETURNS return_data_type ] [ APPEND ONLY ]
```

Comment:
Can we use `result_data_type` here to maintain wording consistency? Or use `return_type` everywhere, just want consistency.